### PR TITLE
feat(metering): implement gas metering and cost allocation system (#190)

### LINKED ISSUE #190: Implement a gas metering and cost allocation system for multi-tenant contract usage
## Description
Build a gas metering system that tracks per-tenant resource consumption, enforces quotas, and enables fair cost allocation across clinics/organizations sharing the contract infrastructure.

## Requirements
- Create `contracts/metering/` with per-address gas tracking
- Implement hierarchical tenant structure: Organization → Clinic → Provider → Patient
- Track gas consumption per operation type (read, write, compute, storage)
- Enforce per-tenant quotas with configurable limits and burst allowances
- Implement cost allocation: distribute contract execution costs proportionally
- Add billing cycle management with usage reports
- Support prepaid and postpaid billing models
- Implement gas token for internal cost settlement
- Add alerts/events when tenants approach quota limits

## Files to Modify
- `contracts/metering/src/lib.rs` (new)
- `contracts/metering/src/quota.rs` (new)
- `contracts/metering/src/billing.rs` (new)
- `contracts/metering/src/gas_token.rs` (new)
- `contracts/vision_records/src/lib.rs` — add metering hooks
- `contracts/common/src/lib.rs` — shared metering types

## Acceptance Criteria
- [ ] Per-tenant gas tracking is accurate
- [ ] Quota enforcement blocks over-limit operations
- [ ] Cost allocation matches actual resource consumption
- [ ] Billing reports are accurate and queryable
- [ ] Gas token settlement works correctly
- [ ] Tests cover quota edge cases and concurrent usage

### PR DESCRIPTION
**Pull request by Tijesunimi004**
Closes #190

---

Add a full per-tenant gas metering, quota enforcement, and cost allocation system for multi-tenant contract usage.

### src/lib.rs
- MeteringContract with initialize, tenant management, quota config, gas recording with hierarchical rollup, billing cycle management, invoice management, and gas token management.
- Tenant hierarchy: Organization → Clinic → Provider → Patient. Consuming gas at a lower level propagates (rolls up) to every ancestor automatically.
- OperationType enum: Read / Write / Compute / Storage.
- GasCosts struct (admin-configurable): read=1, write=5, compute=10, storage=3 by default.
- 80 % alert threshold: QuotaAlertEvent emitted when a tenant crosses 80 % of their total_limit.
- MeteringError covers all failure modes with stable u32 codes.

### src/quota.rs
- TenantQuota: per-type limits (read/write/compute/storage), total cap, burst allowance, enabled flag.
- QuotaUsage: per-type counters plus burst_used.
- check_quota: validates headroom (including burst) without mutating state.
- consume_quota: commits usage; overflow spills into burst_used.
- reset_usage: zeroed at start of each billing cycle.

### src/billing.rs
- BillingModel: Prepaid | Postpaid.
- BillingCycle: open/close lifecycle, timestamps, CycleStatus.
- TenantUsageRecord + BillingReport for cycle-close reporting.
- Invoice: issued for postpaid tenants at cycle close; settable by tenant.
- open_cycle / close_cycle / get_cycle helpers.
- Prepaid balance helpers: credit_prepaid / debit_prepaid.
- Invoice helpers: create_invoice / get_invoice / settle_invoice.

### src/gas_token.rs
- GasTokenAccount snapshot (balance, total_minted, frozen flag).
- mint / burn / freeze / unfreeze / balance_of / total_supply.
- Prepaid burn is called by record_gas before usage is committed.
- Freeze prevents spending; unfreeze restores it.

### src/events.rs
- Rich event set: TenantRegistered, GasRecorded, QuotaAlert, QuotaExceeded, CycleOpened, CycleClosed, InvoiceIssued, InvoiceSettled, GasTokenMinted, GasTokenBurned.

### src/test.rs  (50+ tests)
- Initialisation, gas costs, admin enforcement.
- Tenant registration: all four levels, duplicates, missing parent.
- Quota: set/remove/usage-starts-at-zero.
- Gas recording: all four op types, inactive tenant rejection.
- Quota enforcement: within limit, hard block, burst draw-down, disabled quota, exact boundary.
- Hierarchical rollup: provider→clinic→org, multi-provider aggregation.
- Billing cycle lifecycle: open/close, double-open error, status transitions, usage reset on new cycle, report correctness.
- Postpaid: invoice creation on close, settle, double-settle error, missing invoice.
- Prepaid: mint, burn-on-record, insufficient balance, freeze/unfreeze, total supply tracking, zero-mint error.
- Authorisation: non-admin blocked on all admin operations.
- Multi-cycle cost allocation proportionality.

## Modified: contracts/common/src/metering.rs (new)
- MeteringHook: load/configure/is_configured helpers used by consumer contracts (stored in instance storage under METERING_CONTRACT key).
- MeteringOpType enum mirroring metering::OperationType.

## Modified: contracts/common/src/lib.rs
- Expose metering module and re-export MeteringHook + MeteringOpType.

## Modified: contracts/vision_records/src/lib.rs
- Import MeteringHook, MeteringOpType from teye_common.
- meter_op() internal helper: best-effort fire-and-forget; increments a persistent per-address operation counter so the metering contract can retrieve aggregate stats without blocking the vision record operation.
- configure_metering() public fn (ContractAdmin+ required): sets or clears the metering contract address.
- get_metering_contract() public fn: returns configured address.
- Hooks added in add_record (Write) and get_record on success (Read).

## Modified: Cargo.toml
- Added contracts/metering to workspace members.

## Summary by CodeRabbit

* **New Features**
  * Comprehensive metering and cost-tracking system for resource operations
  * Billing cycle management with prepaid and postpaid models, invoicing and settlement
  * Per-tenant quota enforcement with alerts and limits
  * Gas-token accounting and account controls (mint/burn/freeze)
  * Event emission for metering lifecycle and alerts
  * Metering integration and configurable hook in the records contract

* **Tests**
  * Extensive test suite covering metering, quotas, billing, tokens, and integration

### CODE REVIEW

File: contracts/metering/src/gas_token.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    extend_ttl(env, &key);
+}
+
+/// Full account snapshot for a tenant.
+pub fn get_account(env: &Env, tenant: &Address) -> GasTokenAccount {
+    GasTokenAccount {
+        tenant: tenant.clone(),
+        balance: balance_of(env, tenant),
+        total_minted: env.storage().instance().get(&GT_TOTAL).unwrap_or(0),
+        total_burned: 0, // Aggregate burn tracking is additive; see `burn`.
```
[the remark this answers is not in the document]
Comment (Tijesunimi004) [reply contracts/metering/src/gas_token.rs]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/Stellar-Teye/Teye-Contracts/pull/216#discussion_r2849393046)

File: contracts/metering/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,753 @@
...
+
+        // Compute percentage: (used * 100) / limit
+        let pct = total_used
+            .saturating_mul(100)
+            .checked_div(quota.total_limit)
+            .unwrap_or(0);
+
+        if pct >= ALERT_THRESHOLD_PCT {
+            events::publish_quota_alert(env, tenant.clone(), pct as u32);
+        }
```
[the remark this answers is not in the document]
Comment (Tijesunimi004) [reply contracts/metering/src/lib.rs]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/Stellar-Teye/Teye-Contracts/pull/216#discussion_r2849393078)
